addons/hi-res-thumbnails/userscript.js

```diff
@@ -5,10 +5,29 @@ export default async function ({ addon, console }) {
     });
 
     const lazy = image.classList.contains("lazy");
-    const src = lazy ? image.dataset.original : image.src;
+    let src = lazy ? image.dataset.original : image.src;
 
+    // Don't process data: URLs, since these are never thumbnails to be upsized,
+    // and can be lengthy strings - which really does cause issues when it comes
+    // to the cdn2 regex!
     if (src.startsWith("data:")) continue;
 
+    // If the image is from uploads.scratch.mit.edu, reformat src so it looks
+    // like a cdn2 URL.
+    if (src.includes("uploads.scratch.mit.edu")) {
+      const id = src.match(/[0-9]+/);
+      if (src.includes("projects")) {
+        // Project thumbnails are always 480x360.
+        src = `//cdn2.scratch.mit.edu/get_image/project/${id}_480x360.png`;
+      } else if (src.includes("users")) {
+        // Max user avatar size is 500x500.
+        src = `//cdn2.scratch.mit.edu/get_image/user/${id}_500x500.png`;
+      } else if (src.includes("galleries")) {
+        // Max studio thumbnail size is 500x500.
+        src = `//cdn2.scratch.mit.edu/get_image/gallery/${id}_500x500.png`;
+      }
+    }
+
```

Comment:
I went with just rewriting `src` rather than applying values directly to `width` and `height` values because it makes reconstructing the URL quite a bit easier. This is more processing work than e.g. manually formatting the final width/height/id/type values (as we would need to), but it's minuscule—overall this only needs to run once for each (uploads) image, and at most there are 50 or those present total.